sds_data_manager/lambda_code/SDSCode/batch_starter.py

```diff
@@ -288,160 +183,197 @@ def is_job_in_status_table(
+    if results:
+        return True
     return False
 
 
-def lambda_handler(event: dict, context):
-    """Lambda handler."""
-    logger.info(f"Event: {event}")
-    logger.info(f"Context: {context}")
+def try_to_submit_job(session, batch_client, job_info):
+    """Try to submit a batch job with the given job information.
 
-    # Event details:
-    filename = event["detail"]["object"]["key"]
-    components = ScienceFilePath.extract_filename_components(filename)
-    logger.info(f"Parsed filename - {components}")
-    instrument = components["instrument"]
+    Go through the job information to retrieve all necessary input files
+    (upstream dependencies). If any are missing, return. If we have
+    all the necessary input files, submit the job to the batch queue.
 
-    # Get information for the batch job.
-    region = os.getenv("REGION")
-    account = os.getenv("ACCOUNT")
-    # Create a batch client
-    batch_client = boto3.client("batch")
+    Parameters
+    ----------
+    session : orm session
+        Database session.
+    batch_client : boto3.client
+        Boto3 client for AWS Batch.
+    job_info : dict
+        Dictionary containing components with dates and versions appended.
+
+    Returns
+    -------
+    bool
+        Whether or not this job is ready to be processed.
+    """
+    instrument = job_info["instrument"]
+    data_level = job_info["data_level"]
+    descriptor = job_info["descriptor"]
+    start_date = job_info["start_date"]
+    version = job_info["version"]
+
+    logger.info("Checking for job in progress before looking for dependencies.")
+
+    if is_job_in_processing_table(
+        session=session,
+        instrument=instrument,
+        data_level=data_level,
+        descriptor=descriptor,
+        start_date=start_date,
+        version=version,
+    ):
+        logger.info(
+            f"Job already in progress for {instrument}, {data_level}, "
+            f"{descriptor}, {start_date}, {version}"
+        )
+        return
+
+    # Find the files that this job depends on
+    upstream_dependencies = get_dependencies(
+        session=session,
+        instrument=instrument,
+        data_level=data_level,
+        descriptor=descriptor,
+        direction="UPSTREAM",
+        relationship="HARD",
+    )
+
+    for upstream_dependency in upstream_dependencies:
+        upstream_instrument = upstream_dependency["instrument"]
+        upstream_data_level = upstream_dependency["data_level"]
+        upstream_descriptor = upstream_dependency["descriptor"]
+
+        # Check to see if each upstream dependency file is available
+        # TODO: Update start_date / version request to be more specific
+        #       Currently we are using the same as the job product, but the versions
+        #       may not match exactly if one dependency updates before another
+        upstream_start_date = start_date
+        upstream_version = version
+        upstream_dependency.update(
```

Comment:
I like these changes and it seems like it will make it easier to accommodate weird complicated dependencies